### DISCUSSION TITLE
Sentry Metrics - Now in Beta 🔥🔥🔥

### DISCUSSION THREAD

**Discussion by ale-cota** [getsentry/sentry#64195] [Launch Feedback]
This post is deprecated - please see this discussion on Span Metrics instead which has shipped already: https://github.com/getsentry/sentry/discussions/81239
---- 
Sentry’s latest product is now in Alpha. With metrics you can track and visualize the data points you care about over time, making it easier to monitor your application health and identify issues.

You can create **custom metrics** that matter to your team, like processing time, checkout conversion rate, or user sign ups, and also **explore out-of-the-box metrics**, like transaction duration.

https://github.com/getsentry/sentry/assets/26696515/53c84179-7185-4cc2-9100-0e730456aa6c

If you find something interesting, like a spike in processing time, drill into specific events and use **correlated traces** and other signals like profiles to find the cause of the issue.

https://github.com/getsentry/sentry/assets/26696515/4cc6d8d6-70a5-4a75-93b3-7656d3a58baf

Create metric-based alerts to get notified about unexpected changes or spikes and stay ahead of issues. If you want to continuously monitor a particular metric,  add the widget to a Dashboard for quick comparison and reference. 

<img width="1325" alt="Alerts_Dashboards" src="https://github.com/getsentry/sentry/assets/26696515/c505ee95-8800-49fd-838a-c596aed03e46">

### Sign-up for the waitlist ###

Sentry Metrics is currently in Alpha for our Javascript, Python, PHP, .NET, and Rust SDKs. Features are still in-progress and may have bugs so you’ll likely encounter a few - we recognize the irony.

Join our [waitlist](https://sentry.io/lp/metrics/) to be among one of the first to try it out. You’ll be contacted when it’s your turn via email.

### Let us know what you think ###

If you are an early adopter, we’d love to hear what you think. How are you metrics? How is it helping you and your team? How could it be improved?

Please post down below to ask any questions and let us know your thoughts.

**Comment by ghost** [comment]
This sounds great! But I am a little confused how it differs from the [custom metrics](https://docs.sentry.io/product/performance/metrics/?original_referrer=https%3A%2F%2Fwww.google.com%2F#custom-performance-metrics) already in the product, aside from maybe better dashboarding features?

**Reply by ale-cota** [reply]
Thank you for bringing this up! We recognise that this might be confusing at the moment - definitely something we'll work to improve on in both naming and docs in the short term.

The difference is that [custom performance metrics](https://docs.sentry.io/product/performance/metrics/?original_referrer=https%3A%2F%2Fwww.google.com%2F#custom-performance-metrics) are tied to transactions today and are much more limited. Much like tags they are sent together with the transaction event itself and represent one specific value that you can send. Since transactions are very rich/heavy events, they can also be subject to sampling, so in this case the custom performance metrics are also lost when a transaction is discarded.
[Custom metrics](https://docs.sentry.io/product/metrics/?original_referrer=https%3A%2F%2Fwww.google.com%2F) are "proper metrics", so you can define them as counters, distributions, gauges or sets. They are sent independently of transactions, are much more lightweight and flexible. You also have richer capabilities in terms of creating alerts and dashboards when using the new custom metrics.

When we first introduced the "legacy" custom performance metrics, we didn't have the intention to create a custom metrics product yet and was somewhat of a bridge solution. In the mid-long term, the goal is to consolidate the two. Hope that helps to clarify the distinction, let me know if you have other questions!

**Reply by zaicevas** [reply]
What about sampling for (Alpha) custom metrics? Is it sampled by default and is it possible to adjust?

**Comment by rix501** [comment]
This is super exciting! Are there calculations for metrics? And also, for counter metrics, can we see count?

One thing I would like to do is calculate late sum/count for a specific metric as a success percentage (0 = success, 1 = failure)

**Reply by martinbean** [reply]
Can you not just emit a distribution metric in that case?

```php
\Sentry\metrics()->distribution(
    key: 'foo',
    value: 87.0,
    unit: \Sentry\Metrics\MetricsUnit::percent(),
)
```

**Reply by rix501** [reply]
the issue is I want the percentage across requests

**Comment by SamyPesse** [comment]
One first feedback: The different JS functions can't be used in a cloudflare worker environment where I/O is not allowed outside the request, it fails with:

```
Error: Disallowed operation called within global scope. Asynchronous I/O (ex: fetch() or connect()), setting a timeout, and generating random values are not allowed within global scope. To fix this error, perform this operation within a handler. https://developers.cloudflare.com/workers/runtime-apis/handlers/
```

I'm trying to find a workaround, but it doesn't look like there is one at the moment.

**Reply by ale-cota** [reply]
Thanks a lot for sharing this, Samy! @AbhiPrasad do you have any ideas on how to work around this? 🙏

**Reply by AbhiPrasad** [reply]
Sentry JS SDK does not officially support cloudflare workers yet for metrics - please open a GH issue on the JS SDK repo and we can prioritize accordingly! https://github.com/getsentry/sentry-javascript/issues

**Reply by ale-cota** [reply]
Thanks for the clarification, @AbhiPrasad! Will create the issue.

**Comment by philip-zhan** [comment]
Very exciting stuff! Is there any documentation on the equations currently supported?

**Reply by AbhiPrasad** [reply]
As per https://docs.sentry.io/product/metrics/metrics-explorer-page/ we support the `+`, `-`, `*`, and `/` operators (addition, subtraction, multiplication, division).

**Comment by ale-cota** [comment]
### Update ###
We've now transitioned Sentry Metrics from alpha to beta 🚀
Lots of new features such as equation support, multi-query charts and improvements to plotting and trace correlation. We've also added metrics to Ruby, Unity, Java, Android and React-Native 🔥 
We're excited to hear your feedback!

**Reply by jwarlander** [reply]
Great news! I'd love to try it out, but our primary use cases are in Go code, which isn't yet supported :sweat_smile: Is it possible at the moment to send metrics manually via an HTTP call in a reasonably straightforward way?

**Reply by joshhornby** [reply]
Was just coming here to ask the same as @jwarlander - anyway to send metrics via HTTP would be great.

**Comment by craigbehnke** [comment]
This looks really cool, but I'm curious -- what's the SaaS pricing situation on metrics now/when it goes GA? Will they use Performance Units (if so, how many) or are they included for no extra charge as the docs seem to indicate?

I personally like the free option, but I'm biased.

**Reply by ale-cota** [reply]
Hi Craig,
Thanks for the question! Metrics is currently free while we are in the beta phase, however we will start charging for them after GA (targeting early June). The exact pricing model is not yet finalised, but they will be charged separately from Performance Units.

**Comment by klemmchr** [comment]
Metrics look great so far. The only thing that is a bit confusing to me is the default view that can be saved. This functionality is not scoped to the project but for the user (or for everyone, can't say for sure). This is a bit clunky since switching between projects also retains the metrics displayed. We have projects where metrics vary drastically from each other. For me a functionality like a project scoped default view is missing right now.

**Reply by ale-cota** [reply]
Hi @klemmchr,
Sorry for the delayed response on this and thanks a lot for the feedback. Indeed the default view is scoped to a user, but it does also save a selected project. It is however only one, so I can understand why you might want different ones per project.

Using Dashboards would be a better option in that case (available on Business plans and above). You can save your collection of queries to a dashboard which is shared among teams and can also be scoped to projects.

**Reply by klemmchr** [reply]
This is exactly my issue: without a business plan and dashboards metrics are not really usable right now. I'm not bothered to always recreate my metrics when I want to get quick and easy insights. This is preventing me from using a feature that can be really useful. It's just behind a paywall right now.

**Reply by ale-cota** [reply]
Thanks for sharing that honest feedback. Dashboards in general are a Business feature, but we recognise that for metrics they are especially useful to a larger segment of our customers. We'll continue to gather input on this and maybe reconsider it in the future.

**Comment by george-silva** [comment]
Nice work! Will metrics be available on on-premise installations?

**Reply by JeremiaAu** [reply]
I created an Issue in the self-hosted repo: https://github.com/getsentry/self-hosted/issues/3080

**Comment by Artein** [comment]
This feature looks really interesting. We are checking it out while the feature is in Beta and free.
Anyway, I don't like the restrictions on a metric's name.

1. If the name has '/' or ' ' (space), each character will be replaced with '_' (underscore).
So, in my case when I want to calculate total endpoint price, it is messed with other metrics (my endpoints start with 'api/')

2. If the name of metric is long, it is being cut and only via hovering mouse over it you can read the whole name

3. For development and staging environments would be great to update metrics in UI way quicker than it is now (5 minutes?)

4. If tag name contains ' ' (space), I can't group by that tag

Overall, it looks not that readable atm and should be improved.
@ale-cota, do I miss something and I can improve that on my end?

<img width="449" alt="SCR-20240517-hrmu" src="https://github.com/getsentry/sentry/assets/1109722/2e3bd7ec-f693-4c27-b913-49ffe953fcd4">

**Reply by ale-cota** [reply]
Hi, @Artein! Thanks for the feedback and sorry for the delayed response.
(1) and (4) [Here](https://docs.sentry.io/product/explore/metrics/#supported-characters) you can find more details about supported characters in both metric names, tag keys and tag values. We generally advise not to use spaces.
(2) Yes, we've had to limit the width of that component to optimise for displaying queries all in one row. We hope it's more of an exception that metric names are so long that you have to hover to see the whole name.
(3) We're aware of this delay in first metric ingestion (it should be a couple of minutes) and we will try to improve this in the future.

**Comment by flygerian** [comment]
Hi! Is there a way to maybe send metrics over an HTP request? I migrated my project from python to Go, and I can see that metrics is not supported by the Go SDK, bit metrics are super important to my project currently

**Reply by icebob** [reply]
An HTTP metrics endpoint would be great because it allows to send metrics from any languages, any framework without an SDK.

**Comment by jkytomak** [comment]
Is there way to get same `browser.name `and `geo.country_code` tags as in performance metrics added to custom metrics?

**Reply by ale-cota** [reply]
Hi, @jkytomak! This is not possible today, you would have to retrieve those tags yourself and manually add them to your custom metrics. In the future, however, when metrics will be more tightly coupled with spans (see [this comment](https://github.com/getsentry/sentry/discussions/64195#discussioncomment-9762049) below) this should be more easily possible.

**Comment by cezarsmpio** [comment]
I have some piece of feedback:

- I liked the option of "creating one chart per metric" but I'd like to rename those charts instead of having "avg(my_custom_metric)", I'd like to have a more meaningful name 
- Custom metrics are not reset from the metrics dropdown when I switch projects.

**Reply by ale-cota** [reply]
Thanks for the feedback, @cezarsmpio! 🙏 we'll try to improve this.

**Comment by rafalmemyard** [comment]
This is a great addition to the Sentry service, but I'm having trouble with visibility of metrics (or delays in flushing metrics) using Java SDK v7.9.0.  It looks like sometimes it takes hours for the metrics to show up in the Sentry dashboards, and sometimes I'm not sure they show up at all (which makes for rather poor metrics, since I can't trust that they'll get there).  Should I file a separate ticket?

**Reply by ale-cota** [reply]
Hi, @rafalmemyard! Sorry for the delayed reponse on this. We are aware that there can be some delay in metrics ingestion, but this should be more in the order of minutes and not hours. Would you mind reaching out on [Discord](https://discord.gg/sentry) or [our support](https://sentry.zendesk.com/hc/en-us/requests/new) to share more details about your project and configuration? Thanks a lot!

**Reply by rafalmemyard** [reply]
7.11.0 fixed this for me.

**Comment by JeremiaAu** [comment]
asp.net projects are listed under "unsupported" even though it works:
 
![image](https://github.com/getsentry/sentry/assets/96953097/fbdc4e6d-f2e5-4190-81aa-53740789b130)
(I am using self-hosted 24.5.0)

**Reply by ale-cota** [reply]
Thanks for reporting, @JeremiaAu 🙏 we'll fix this.

**Reply by ale-cota** [reply]
This has now been fixed

**Comment by gopi-skuvent** [comment]
Hi, I'm using Node (and Express) and already use Sentry for issues, performance and other features. Now trying out Metrics and I've added this code in one of our APIs and after calling the API several times, no custom metrics shows up in the Sentry UI. I've waited more than an hour and see nothing. What else should I do? I've went through this [Sentry Metrics documentation](https://docs.sentry.io/platforms/javascript/guides/node/metrics/). 


```javascript
Sentry.metrics.increment('item.product', 1, {
 tags: { 'item.status': 'active' },
});
```

**Reply by gopi-skuvent** [reply]
Ah, just after posting this, I re-read the documentation (duh, RTFM), I had to enable experiment flag in `Sentry.init` like this. 

```javascript
_experiments: {
  metricsAggregator: true,
},
```

**Comment by ale-cota** [comment]
## [Announcement] Upcoming API changes

Hi everyone,

Through the beta process so far, we've learned that metrics that are connected to traces give our users a more intuitive and connected debugging workflow. We're committed to surfacing actionable insights in Sentry and we think tracing is core to helping developers find the root cause of an unexpected behaviour in their app.

To deliver a more effective metrics experience, we will be transitioning to a new Metrics API that ensures all collected metrics have a correlated span. You'll eventually need to migrate to the new APIs once released and we will continue to support the existing for a sufficient period of time to allow for a smooth transition. 

We believe that this change will enhance debugging capabilities by ensuring metrics are consistently linked with all your Sentry data through traces. We hope you agree and encourage you to continue providing feedback, so we can deliver the best possible product experience.

For more context and details about this upcoming change, please check out [this article](https://sentry.zendesk.com/hc/en-us/articles/26369339769883-Upcoming-API-Changes-to-Metrics).

**Reply by dirkdev98** [reply]
Liking the metrics so far, to give insight in the non-sampled view of what is happening. Which allows for creating alerts based on real expected usage..

Although the expected solution is to 'reconstruct a fairly accurate representation of your data', I am not sure if this would cover all my use cases. We filter out spans, or don't even start them in some paths, where we still like to have metrics.
For example, a case we currently cover with metrics, is to keep track of how often a poll-based queue is waiting on a job. This way we can detect over- / under-provisioning. Only if a job is found, we start a span, which is then either sampled or not. 

We could keep track of this value separately from Sentry, detect if a sampled span is started and then attach the metrics, but would like if the SDK does that for us :)

**Reply by mramaiya** [reply]
If we have a metric that is, for example, count of pending db queries, how would that be connected to a span?

The use I have for this, whether or not that is right, is to monitor overall application health. 

These metrics would specifically not be connected to an individual span. 

Is that ability going to get lost with this transition?

**Reply by ale-cota** [reply]
@mramaiya thanks for the feedback and the question. No, the ability itself will not be lost, but indeed this approach might be more immediately useful for some use cases than others. You will still be able to collect a metric for the count of pending db queries and have the same in-product functionality as today, but the measurements do need to be attached to a span in order to send them - whether it's an existing one or a custom span that you create. 

The idea of always connecting metrics to spans is to offer useful information for debugging that you can get through tracing, so you can do more than just observe them. We believe in a large majority of cases having the extra context that comes through tracing can help to understand what was going on in the system when the metric behaved unexpectedly.

**Reply by mramaiya** [reply]
So, if I understand correctly, the metric count will still be global, but inc/dec operation will be tied to a span?

**Reply by jtomaszewski** [reply]
So I guess we'll have to create some "global span" that will be alive "forever", and then such "global metrics" would be attached to such a span.

Is that how you want us to have it? I hope SDK will make that somehow easy.

**Reply by tjerkw** [reply]
Personally having them decoupled makes more sense for the reasons mentioned above. Most custom metrics are not span related, but more operationally system related: Total count of event x for example. 

Why not use OpenTelemetry based metrics collection?

**Reply by filips123** [reply]
What about providing two APIs for metrics, one coupled with spans and one not? There are good use cases for both approaches, and letting the developer choose whether they want a metric to be connected with a span or not seems like a good solution for all use cases.

**Comment by gopi-skuvent** [comment]
Hi Alexandra,

I missed setting "metricsAggregator" to "true". After that, the metrics
started flowing. So we can close this ticket.

Thanks,
Gopi


On Thu, Jun 13, 2024 at 2:41 AM Alexandra Cota ***@***.***>
wrote:

> Thanks for reporting, @JeremiaAu <https://github.com/JeremiaAu> 🙏 we'll
> fix this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/getsentry/sentry/discussions/64195#discussioncomment-9761439>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BDTHJK3YCNEW6HUQX2DX3FTZHFSN7AVCNFSM6AAAAABCRYIVUSVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TONRRGQZTS>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Gopi Sundharam
*Skuvent, Inc* | Co-Founder, CTO
***@***.*** | m: 408-836-0944

**Reply by ale-cota** [reply]
Hi Gopi,
Glad it's working now and thanks for letting us know :)

**Comment by pecpwee** [comment]
I am using Metrics to track the success rate of network requests on my Android App. 
I calculated the success rate by using the "add equation" method on the Metrics page, which divides the number of successful requests by the total number of requests to get the percentage success rate. 
However, I am unable to set alerts based on this percentage metric. 
For example, I would like to be able to receive an alert if the network request success rate falls below 90%.this functionality is not available In the current situation,

Can this functionality be added?

Looking forward to your reply, thank you! @ale-cota 
![image](https://github.com/user-attachments/assets/a75c524a-e3ef-42f9-8a79-852e9565164a)

**Comment by agentreno** [comment]
Hello, we're building a Cordova mobile app but metrics are not yet supported in the Cordova SDK. So I'm trying to use a second browser SDK instance where metrics are supported, but keep the instances separate as per the guidance here:
https://docs.sentry.io/platforms/javascript/best-practices/multiple-sentry-instances/

I see in those docs this approach isn't recommended but we're stuck between a rock and a hard place because although the Cordova SDK is missing metrics support, it still has some functionality we need around native crashes, offline storage etc.

Because the `metrics` module is on the Sentry global, I don't think I'm able to use a separate client/scope etc. as described in the multiple instances docs. Is what I'm trying to do even possible right now?

**Reply by vctormb** [reply]
I'd like to know that as well. I'm using Sentry in a shared environment with a new scope and I'd like to use metrics too.

**Comment by helmut-hoffer-von-ankershoffen** [comment]
How to get metrics out of Sentry? Is there a Rest API or Prometheus exporter?
